BoxContentSDK/BoxContentSDK/Requests/BOXBookmarkUnshareRequest.m

```diff
@@ -60,14 +60,24 @@ - (void)performRequestWithCompletion:(BOXBookmarkBlock)completionBlock
     BOXAPIJSONOperation *bookmarkOperation = (BOXAPIJSONOperation *)self.operation;
 
     bookmarkOperation.success = ^(NSURLRequest *request, NSHTTPURLResponse *response, NSDictionary *JSONDictionary) {
+        BOXBookmark *bookmark = [[BOXBookmark alloc] initWithJSON:JSONDictionary];
+        if ([self.cacheClient respondsToSelector:@selector(cacheBookmarkUnshareRequest:withBookmark:error:)]) {
+            [self.cacheClient cacheBookmarkUnshareRequest:self
+                                             withBookmark:bookmark
+                                                    error:nil];
+        }
         if (completionBlock) {
-            BOXBookmark *bookmark = [[BOXBookmark alloc] initWithJSON:JSONDictionary];
             [BOXDispatchHelper callCompletionBlock:^{
                 completionBlock(bookmark, nil);
             } onMainThread:isMainThread];
         }
     };
     bookmarkOperation.failure = ^(NSURLRequest *request, NSHTTPURLResponse *response, NSError *error, NSDictionary *JSONDictionary) {
+        if ([self.cacheClient respondsToSelector:@selector(cacheBookmarkUnshareRequest:withBookmark:error:)]) {
+            [self.cacheClient cacheBookmarkUnshareRequest:self
+                                             withBookmark:nil
+                                                    error:error];
+        }
```

Comment:
It seems like in general the code is very duplicative. L76-89 are just like L63-68, if a `bookmark` and `error` ivars were initialized to `nil` or parameterized in a helper method.

What leads to these sorts of patterns? Is it the nature of the language (Obj-C) or that nature of the iOS frameworks? Or is it the patterns we (Box) have developed?

Reply:
I think it mainly stems from wanting different params to success and failure blocks, because Objective-C can't express sum types. I'd love to see a different pattern in non-objective-c-supporting Swift. I agree that this could be abstracted internally here to reduce duplication.

Reply:
I think the main reason is for clarity. There are some cases we would need different logic in success and failure block. Maybe cleaner to have the shared block if we need to extend the shared logic further.